service/worker/archiver/client.go

```diff
@@ -75,20 +87,67 @@ func NewClient(
 	publicClient workflowserviceclient.Interface,
 	numWorkflows dynamicconfig.IntPropertyFn,
 	requestRPS dynamicconfig.IntPropertyFn,
+	archiverProvider provider.ArchiverProvider,
 ) Client {
 	return &client{
-		metricsClient: metricsClient,
-		logger:        logger,
-		cadenceClient: cclient.NewClient(publicClient, common.SystemLocalDomainName, &cclient.Options{}),
-		numWorkflows:  numWorkflows,
-		rateLimiter:   tokenbucket.NewDynamicTokenBucket(requestRPS, clock.NewRealTimeSource()),
+		metricsClient:    metricsClient,
+		logger:           logger,
+		cadenceClient:    cclient.NewClient(publicClient, common.SystemLocalDomainName, &cclient.Options{}),
+		numWorkflows:     numWorkflows,
+		rateLimiter:      tokenbucket.NewDynamicTokenBucket(requestRPS, clock.NewRealTimeSource()),
+		archiverProvider: archiverProvider,
 	}
 }
 
 // Archive starts an archival task
-func (c *client) Archive(request *ArchiveRequest) error {
+func (c *client) Archive(ctx context.Context, request *ClientRequest) error {
 	c.metricsClient.IncCounter(metrics.ArchiverClientScope, metrics.CadenceRequests)
 
+	taggedLogger := tagLoggerWithRequest(c.logger, *request.ArchiveRequest).WithTags(
+		tag.ArchivalCallerServiceName(request.CallerService),
+		tag.ArchivalArchiveInline(request.ArchiveInline),
+	)
+	if request.ArchiveInline {
+		return c.archiveInline(ctx, request, taggedLogger)
+	}
+
+	return c.sendArchiveSignal(ctx, request.ArchiveRequest, taggedLogger)
+}
+
+func (c *client) archiveInline(ctx context.Context, request *ClientRequest, taggedLogger log.Logger) (err error) {
+	defer func() {
+		if err != nil {
+			c.metricsClient.IncCounter(metrics.ArchiverClientScope, metrics.ArchiverClientInlineArchiveFailureCount)
+			taggedLogger.Error("failed to perform workflow history archival inline", tag.Error(err))
+			err = c.sendArchiveSignal(ctx, request.ArchiveRequest, taggedLogger)
```

Comment:
Can we call this in Archive method. Calling signal in the inline method feels like the wrong level of abstraction to me. These two methods should simply know how to do their respective types of archival - the flow control belongs one level higher.

Reply:
Makes sense. Updated.